File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/CompiledLibraryMultiResults.java
package org.cqframework.cql.cql2elm;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import org.cqframework.cql.cql2elm.model.CompiledLibrary;
import org.hl7.elm.r1.Library;
import org.hl7.elm.r1.VersionedIdentifier;

/**
 * Represents the result of compiling a CQL library, including the compiled library and any compilation errors, in
 * order to support a partially successful compile with errors
 */
// LUKETODO:  record?
public class CompiledLibraryMultiResults {
    private final List<CompiledLibraryResult> results;

    public static CompiledLibraryMultiResults from(List<CompiledLibraryResult> results) {
        return new CompiledLibraryMultiResults(results);
    }

    private CompiledLibraryMultiResults(List<CompiledLibraryResult> results) {
        this.results = Collections.unmodifiableList(results);
    }

    public boolean hasErrors() {
        return results.stream().anyMatch(res -> !res.errors().isEmpty());
    }

    public List<CompiledLibraryResult> allResults() {
        return results;
    }

    public List<CompiledLibrary> allCompiledLibraries() {
        return results.stream().map(CompiledLibraryResult::compiledLibrary).toList();
    }

    public CompiledLibraryResult getOnlyResult() {
        if (results.size() != 1) {
            throw new IllegalStateException("Expected exactly one result, but found " + results.size());
        }
        return results.get(0);
    }

    public List<CqlCompilerException> getErrorsFor(VersionedIdentifier libraryIdentifier) {
        return results.stream()
                .filter(res -> libraryIdentifier.equals(res.compiledLibrary().getIdentifier()))
                .map(CompiledLibraryResult::errors)
                .flatMap(Collection::stream)
                .toList();
    }

    public List<Library> allLibrariesWithoutErrorSeverity() {
        return results.stream()
                .filter(res -> !CqlCompilerException.hasErrors(res.errors()))
                .map(CompiledLibraryResult::compiledLibrary)
                .map(CompiledLibrary::getLibrary)
                .toList();
    }

    public List<CqlCompilerException> allErrors() {
        return results.stream()
                .map(CompiledLibraryResult::errors)
                .flatMap(Collection::stream)
                .toList();
    }
}

File: Src/java/cql-to-elm/src/commonMain/kotlin/org/cqframework/cql/cql2elm/CompiledLibraryMultiResults.kt
package org.cqframework.cql.cql2elm

import org.cqframework.cql.cql2elm.model.CompiledLibrary
import org.hl7.elm.r1.Library
import org.hl7.elm.r1.VersionedIdentifier

/**
 * Represents the result of compiling a CQL library, including the compiled library and any
 * compilation errors, in order to support a partially successful compile with errors
 */
class CompiledLibraryMultiResults(val results: List<CompiledLibraryResult>) {
    fun hasErrors(): Boolean {
        return results.any({ res -> !res.errors.isEmpty() })
    }

    fun allResults(): List<CompiledLibraryResult> {
        return results
    }

    fun getCompiledLibraryFor(libraryIdentifier: VersionedIdentifier): CompiledLibrary? {
        return results
            .firstOrNull { res -> libraryIdentifier.id.equals(res.compiledLibrary.identifier!!.id) }
            ?.compiledLibrary
    }

    fun allCompiledLibraries(): List<CompiledLibrary> {
        return results.map { it.compiledLibrary }
    }

    val onlyResult: CompiledLibraryResult
        get() {
            check(results.size == 1) { "Expected exactly one result, but found " + results.size }
            return results.get(0)
        }

    fun getErrorsFor(libraryIdentifier: VersionedIdentifier): List<CqlCompilerException> {
        return results
            .filter({ res -> libraryIdentifier.equals(res.compiledLibrary.identifier) })
            .flatMap { it.errors }
    }

    fun allLibrariesWithoutErrorSeverity(): List<Library?> {
        return results.filter({ res -> !CqlCompilerException.hasErrors(res.errors) }).map {
            it.compiledLibrary.library
        }
    }

    fun allErrors(): List<CqlCompilerException> {
        return results.flatMap { it.errors }
    }

    companion object {
        fun from(results: List<CompiledLibraryResult>): CompiledLibraryMultiResults {
            return CompiledLibraryMultiResults(results)
        }
    }
}
